runtime/ghex_comm/run.cpp

```diff
@@ -311,14 +328,51 @@ result grid::collect_results(result const &r) const {
 }
 
 void runtime_register_options(ghex_comm, options &options) {
+  options("cart-dims", "dimensons of cartesian communicator", "PX PY", 2);
   options("sub-domains",
           "number of sub-domains (each sub-domain computation runs in its own "
           "thread)",
           "S", {1});
+  options("thread-cart-dims",
+          "dimensons of cartesian decomposition "
+          "among sub-domains",
+          "TX TY", 2);
+#ifdef __CUDACC__
+  options("device-mapping",
+          "node device mapping: device id per sub-domain in the format "
+          "I_0:I_1:...:I_(N-1) "
+          "where I_i are cuda device ids "
+          "and N = #ranks-per-node x S",
+          "M");
+#endif
 }
 
 runtime runtime_init(ghex_comm, options_values const &options) {
-  return runtime(options.get<int>("sub-domains"));
+  std::array<int, 2> cart_dims = {0, 0};
+  if (options.has("cart-dims"))
+    cart_dims = options.get<std::array<int, 2>>("cart-dims");
+  std::array<int, 2> thread_cart_dims = {0, 0};
+  if (options.has("thread-cart-dims"))
+    thread_cart_dims = options.get<std::array<int, 2>>("thread-cart-dims");
+#ifdef __CUDACC__
+  std::vector<int> device_mapping;
+  if (options.has("device-mapping")) {
+    auto s = options.get<std::string>("device-mapping");
+    const std::string delimiter = ":";
+    if (s.back() != ':')
+      s.push_back(':');
+    std::size_t pos = 0;
+    while ((pos = s.find(delimiter)) != std::string::npos) {
+      if (pos > 0u)
+        device_mapping.push_back(std::abs(std::stoi(s.substr(0, pos))));
+      s.erase(0, pos + delimiter.length());
+    }
```

Comment:
I would prefer some more STL usage over manual splitting here, e.g. something like below + throwing an error if a device ID is negative instead of taking the absolute.
```
std::regex re(":");
std::transform(std::sregex_token_iterator(input.begin(), input.end(), re, -1),
               std::sregex_token_iterator(),
               std::back_inserter(device_mapping),
               [](std::string const& token) {
                   int n = std::stoi(token);
                   // throw error here if n is negative
                   return n;
               });
```